src/targets/gpu/jit/mlir.cpp

```diff
@@ -21,39 +21,196 @@
  * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  * THE SOFTWARE.
  */
-#include <migraphx/gpu/compiler.hpp>
+
+#include <iterator>
+#include <migraphx/builtin.hpp>
+#include <migraphx/instruction_ref.hpp>
+#include <migraphx/iterator_for.hpp>
 #include <migraphx/make_op.hpp>
+#include <migraphx/module.hpp>
+#include <migraphx/gpu/compiler.hpp>
 #include <migraphx/gpu/context.hpp>
+#include <migraphx/gpu/code_object_op.hpp>
 #include <migraphx/gpu/mlir.hpp>
+#include <migraphx/gpu/compile_pointwise.hpp>
 
 namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 namespace gpu {
 
+static migraphx::module create_pointwise_module(migraphx::module_ref in_mod)
+{
+    migraphx::module m;
+    std::unordered_map<instruction_ref, instruction_ref> ins_map;
+    for(const auto i : iterator_for(*in_mod))
+    {
+        if(i->name() == "@param")
+        {
+            ins_map[i] = m.add_parameter(any_cast<builtin::param>(i->get_operator()).parameter,
+                                         shape{i->get_shape().type()});
+        }
+        else if(i->name() == "@literal")
+        {
+            ins_map[i] = m.add_literal(i->get_literal());
+        }
+        else if(i->name() == "multibroadcast" and i->inputs().front()->name() == "@literal")
+        {
+            ins_map[i] = ins_map[i->inputs().front()];
+        }
+        else
+        {
+            std::vector<instruction_ref> inputs;
+            inputs.resize(i->inputs().size());
+            std::transform(i->inputs().begin(),
+                           i->inputs().end(),
+                           inputs.begin(),
+                           [&](const auto input) { return ins_map[input]; });
+            if(i->name() == "@return")
+            {
+                m.add_return(inputs);
+            }
+            else
+            {
+                ins_map[i] = m.add_instruction(i->get_operator(), inputs);
+            }
+        }
+    }
```

Comment:
Why doesnt this just do `return *in_mod`?

Reply:
Need to eliminate multibroadcasts

Reply:
Ah I missed that. You can use `add_instructions` with a custom `inserter` function to eliminate broadcasts, but you need to insert the parameters first though:

```cpp
result.add_instructions(
      in_mod, &map_ins,
      [](module &m, instruction_ref ins, const operation &op,
         const std::vector<instruction_ref> &inputs,
         const std::vector<module_ref> &mod_args) -> instruction_ref {
        if (op.name() == "multibroadcast" and
            inputs.front()->name() == "@literal")
          return inputs.front();
        else
          return m.insert_instruction(ins, op, inputs, mod_args);
      });
```

Reply:
In order to use `add_instructions`, i need to first create two lists. 
1. instructions that are parameters 
2. instructions that are not parameters.

Step 1. I would need to iterate the the split pointwise module to get that first. 

Step 2. I would need to insert parameters first and maintain a map of old parameter to new parameter. 

Step 3. Go through second list and all remaining instructions while eliminating broadcasts. This would make use of `add_instructions`. 

Instead of going through three steps i find it better to just make use of `create_pointwise_module()` function, which  combines all those three steps into one. 

Instead of going through

Reply:
You dont need to two lists, you just need a list of parameters which you can get with `get_parameters` and then insert them and then insert the module:

```cpp
static module create_pointwise_module(module_ref in_mod) {
  module result;
  std::unordered_map<instruction_ref, instruction_ref> map_ins;
  for (auto param : in_mod->get_parameters())
  {
    map_ins[param] = m.add_parameter(
        any_cast<builtin::param>(param->get_operator()).parameter,
        shape{param->get_shape().type()});
  }
  result.add_instructions(
      in_mod, &map_ins,
      [](module &m, instruction_ref ins, const operation &op,
         const std::vector<instruction_ref> &inputs,
         const std::vector<module_ref> &mod_args) -> instruction_ref {
        if (op.name() == "multibroadcast" and
            inputs.front()->name() == "@literal")
          return inputs.front();
        else
          return m.insert_instruction(ins, op, inputs, mod_args);
      });
  return result;
}
```

Reply:
`add_instructions` will add parameters again in that case 
https://github.com/ROCm/AMDMIGraphX/blob/2bdd02d38c41802f5883211ffc7d9020b1b9d408/src/module.cpp#L227

and update the `map_ins` map. Therefore i don't think it will work. I need to maintain two lists. 1. parameters and 2. non-parameters instructions and use this overload 
https://github.com/ROCm/AMDMIGraphX/blob/2bdd02d38c41802f5883211ffc7d9020b1b9d408/src/module.cpp#L424

Reply:
Done.